---
title: "restfulSE -- experiments with HDF5 server content wrapped in SummarizedExperiment"
author: "Vincent J. Carey, stvjc at channing.harvard.edu, Shweta Gopaulakrishnan, reshg at channing.harvard.edu, Samuela Pollack, spollack at jimmy.harvard.edu"
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{restfulSE -- experiments with SE interface to remote HDF5}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::pdf_document:
    toc: yes
    number_sections: yes
  BiocStyle::html_document:
    highlight: pygments
    number_sections: yes
    theme: united
    toc: yes
---

# restfulSE

This R package includes proof-of-concept code illustrating several approaches to SummarizedExperiment design with
assays stored out-of-memory.

## HDF5 server backed SummarizedExperiment

[HDF Server](https://support.hdfgroup.org/projects/hdfserver/) "extends the HDF5 data model to efficiently store large data objects (e.g. up to multi-TB data arrays) and access them over the web using a RESTful API."  In this `restfulSE` package,
several data structures are introduced 

- to model the server data architecture and 
- to perform targeted extraction of numerical data from HDF5 arrays stored on the server. 

We maintain, thanks to a grant from the National Cancer Institute,
the server [http://54.174.163.77:5000/](http://54.174.163.77:5000/).  Visit this URL to get a flavor of the
server structure: datasets, groups, and datatypes are high-level elements to be manipulated to work
with data values from the server.

# Some details

## Motivation

Extensive human and computational effort is expended on
downloading and managing large genomic data at site of analysis.
Interoperable formats that are accessible via generic operations
like those in RESTful APIs may help to improve cost-effectiveness
of genome-scale analyses.

In this report we examine the use of HDF5 server as a back end
for assay data, mediated through the RangedSummarizedExperiment API
for interactive use.

A modest server configured to deliver HDF5 content via a
RESTful API has been prepared and is used in this vignette.

## Executive summary

We want to provide rapid access to array-like data.  We'll work
with the Banovich 450k data as there is a simple check against
an in-memory representation.

```{r setup,echo=FALSE}
suppressPackageStartupMessages({
library(restfulSE)
library(SummarizedExperiment)
library(Rtsne)
})
```{r dobanoyy}
library(restfulSE)
bigec2 = H5S_source("http://54.174.163.77:5000")
bigec2
dsmeta(bigec2)[1:2,] # two groups
dsmeta(bigec2)[1,2][[1]] # all dataset candidates in group 1
```

We use double-bracket subscripting to grab a reference to a
dataset from an H5S source.
```{r doba2}
banref = bigec2[["assays"]] # arbitrary name assigned long ago
banref
```

We build a RESTfulSummarizedExperiment by combining an assay-free
RangedSummarizedExperiment with this reference.
```{r doba3}
data(banoSEMeta)
rbano = RESTfulSummarizedExperiment(banoSEMeta, banref)
rbano
```

We can update the SummarizedExperiment metadata 
through subsetting operations, and then extract the relevant
assay data.  The data are retrieved from the remote server
with the `assay` method.
```{r doba4}
rbanoSub = rbano[5:8, c(3:9, 40:50)] 
assay(rbanoSub) 
```

## 10xGenomics examples

### t-SNE for a set of genes annotated to hippocampus

We have used Martin Morgan's TENxGenomics package
to create a dense HDF5 representation of the 
assay data, and placed it on the `bigec2` server.
The metadata are available as `st100k` in this package;
we have used EnsDb.Mmusculus.v79 to supply gene ranges
where available; genes reported but without addresses
are addressed at chr1:2 with width 0.  The rows
are sorted by genomic address within chromosomes.

```{r gettx}
txdat = bigec2[["tenx_100k_sorted"]]
data(st100k)
tenx100k = RESTfulSummarizedExperiment( st100k, 
   txdat )
tenx100k
```

We will subset genes annotated to hippocampus development.
Here are some related categories:

```
12092 GO:0021766                      hippocampus development
12096 GO:0021770            parahippocampal gyrus development
34609 GO:0097410      hippocampal interneuron differentiation
34631 GO:0097432 hippocampal pyramidal neuron differentiation
34656 GO:0097457                      hippocampal mossy fiber
35169 GO:0098686       hippocampal mossy fiber to CA3 synapse
42398 GO:1990026            hippocampal mossy fiber expansion
```

```{r anno}
library(org.Mm.eg.db)
atab = select(org.Mm.eg.db, keys="GO:0021766", keytype="GO", columns="ENSEMBL")
hg = atab[,"ENSEMBL"]
length(hgok <- intersect(hg, rownames(tenx100k)))
```
This is a very scattered collection of rows in the matrix.  We
acquire expression measures for genes annotated to
hippocampus on 4000 samples.  t-SNE is then used to project
the log-transformed measures to the plane.
```{r getdat, cache=TRUE}
hipn = assay(tenx100k[hgok,1:4000])  # slow
d = dist(t(log(1+hipn)), method="manhattan")
proj = Rtsne(d)
```
```{r plt,fig=TRUE}
plot(proj$Y)
```

### A set of genes related to the visual cortex

Tasic et al. (Nature neuro 2016, DOI 10.1038/nn.4216)
describe single cell analysis of the adult murine
brain, identify clusters of cells with distinct
transcriptional profiles and anatomic location,
and enumerate lists of genes that discriminate
these clusters.  The tasicST6 DataFrame provides
details.

```{r lktas}
data(tasicST6)
tasicST6
```

Key high-level discrimination concerns cells regarded as
GABAergic vs. glutamatergic (inhibitory vs excitatory
neurotransmission).


   

## Background

Banovich et al. published a subset of DNA methylation measures
assembled on 64 samples of immortalized B-cells from the YRI HapMap cohort.

```{r lkd}
library(restfulSE)
data(banoSEMeta)
banoSEMeta
```

The numerical data have been exported using H. Pages' 
saveHDF5SummarizedExperiment applied to the banovichSE
SummarizedExperiment in the yriMulti package.  The HDF5
component is simply copied into the server data space
on the remote server.  


## Hierarchy of server resources

### Server

Given the URL of a server running HDF5 server, we create
an instance of `H5S_source`:
```{r doso}
mys = H5S_source(serverURL="http://54.174.163.77:5000")
mys
```

### Groups

The server identifies a collection of 'groups'.  For the server
we are working with, only one group, at the root, is of interest.
```{r lkg}
groups(mys)
```

### Links for a group

There is a class to hold the link set for any group:
```{r lkhl}
lin1 = restfulSE::links(mys,1)
lin1
```

